# codes
IN this repository you find codes for analyzing data and visualizing the results. Information for used softwares can be found at below links.
- [fastp](https://github.com/OpenGene/fastp)
- [FastQC](https://www.bioinformatics.babraham.ac.uk/projects/fastqc/)
- [bwa-mem2](https://github.com/bwa-mem2/bwa-mem2)
- [elPrep](https://github.com/ExaScience/elprep)
- [Samtools](http://www.htslib.org)
- [Bcftools](http://samtools.github.io/bcftools/)
- [popoolation](https://sourceforge.net/p/popoolation/wiki/Main/)
- [popoolation2](https://sourceforge.net/p/popoolation2/wiki/Main/)
- [CAFE](https://github.com/ymat2/crayfish_cafe_analysis)
- [vcf2phylip](https://github.com/edgardomortiz/vcf2phylip)
- [HISAT2](http://daehwankimlab.github.io/hisat2/)
- [StringTie](https://github.com/gpertea/stringtie/tree/master)
- [Trinity](https://github.com/trinityrnaseq/trinityrnaseq/wiki)
- [TCC](https://bioconductor.org/packages/release/bioc/html/TCC.html)

## analysis
Codes used for population genomic and transcriptomic analysis are stored. The `prepDE.py` used for making count matrix is attached to the `StringTie` software and available [here](https://github.com/gpertea/stringtie/blob/master/prepDE.py).

## figures
R codes used for analyzing and visualizing the results.
